sw/device/tests/dif_gpio/dif_gpio_test.cc

```diff
@@ -0,0 +1,407 @@
+                                uint32_t const index) -> bool {
+      return DifGpioTest::mock->read_register_bit(gpio, offset, index);
+    };
+    write_register_fptr = [](dif_gpio_t const *const gpio,
+                             uint32_t const offset, uint32_t const val) {
+      DifGpioTest::mock->write_register(gpio, offset, val);
+    };
+    write_register_masked_access_fptr =
+        [](dif_gpio_t const *const gpio, uint32_t const lowerOffset,
+           uint32_t const upperOffset, uint32_t const mask,
+           uint32_t const val) {
+          DifGpioTest::mock->write_register_masked_access(
+              gpio, lowerOffset, upperOffset, mask, val);
+        };
+    write_register_bit_masked_access_fptr =
+        [](dif_gpio_t const *const gpio, uint32_t const lowerOffset,
+           uint32_t const upperOffset, uint32_t const index, bool const val) {
+          DifGpioTest::mock->write_register_bit_masked_access(
+              gpio, lowerOffset, upperOffset, index, val);
+        };
+  }
+
+  // Helper methods to reduce repetition
+  auto &EXPECT_INDEX_TO_MASK(uint32_t const index) {
+    return EXPECT_CALL(*mock, index_to_mask(index));
+  }
+
+  auto &EXPECT_ABORT_IF_INVALID() {
+    return EXPECT_CALL(*mock, abort_if_invalid(&gpio));
+  }
+
+  auto &EXPECT_REG32_WRITE(uint32_t const offset, uint32_t const val) {
+    return EXPECT_CALL(*mock, reg32_write(gpio.base_addr, offset, val));
+  }
+
+  auto &EXPECT_REG32_READ(uint32_t const offset) {
+    return EXPECT_CALL(*mock, reg32_read(gpio.base_addr, offset));
+  }
+
+  auto &EXPECT_READ_REGISTER(uint32_t const offset) {
+    return EXPECT_CALL(*mock, read_register(&gpio, offset));
+  }
+
+  auto &EXPECT_READ_REGISTER_BIT(uint32_t const offset, uint32_t const index) {
+    return EXPECT_CALL(*mock, read_register_bit(&gpio, offset, index));
+  }
+
+  auto &EXPECT_WRITE_REGISTER(uint32_t const offset, uint32_t const val) {
+    return EXPECT_CALL(*mock, write_register(&gpio, offset, val));
+  }
+
+  auto &EXPECT_WRITE_REGISTER_MASKED_ACCESS(uint32_t const lowerOffset,
+                                            uint32_t const upperOffset,
+                                            uint32_t const mask,
+                                            uint32_t const val) {
+    return EXPECT_CALL(*mock, write_register_masked_access(
+                                  &gpio, lowerOffset, upperOffset, mask, val));
+  }
+
+  auto &EXPECT_WRITE_REGISTER_BIT_MASKED_ACCESS(uint32_t const lowerOffset,
+                                                uint32_t const upperOffset,
+                                                uint32_t const index,
+                                                bool const val) {
+    return EXPECT_CALL(*mock, write_register_bit_masked_access(
+                                  &gpio, lowerOffset, upperOffset, index, val));
+  }
+
+  void TearDown() override {
+    if (mock) {
+      mock.reset();
+    }
+  }
```

Comment:
Don't bother; this will get taken care of during class destruction.

Reply:
Not really, since `DifGpioTest::mock` is `static`, it must be destroyed in `DifGpioTest::TearDown()` to verify its expectations. Otherwise, it is destroyed at program exit and its expectations are not verified. Added a comment. Also, this is the error message you get if you the mock object is leaked:

> ERROR: 1 leaked mock object found at program exit. Expectations on a mock object is verified when the object is destructed. Leaking a mock means that its expectations aren't verified, which is usually a test bug. If you really intend to leak a mock, you can suppress this error using testing::Mock::AllowLeak(mock_object), or you may use a fake or stub instead of a mock.